### PR TITLE
feat: preliminary support for hstore backend

### PR DESCRIPTION
**Pull request by VGalaxies**
## Purpose of the PR

subtask of https://github.com/apache/incubator-hugegraph/issues/2265

Introduce `hugegraph-hstore` module to connect with `hugegraph-pd` and `hugegraph-store`, and make adjustments to other modules as well.

This PR is quite big, and some commits related to maven configuration fixes have already been cherry-picked into https://github.com/apache/incubator-hugegraph/pull/2289. It might be advisable to further split this PR to facilitate the review. (How?)

## Main Changes

- maven configuration fixes (see https://github.com/apache/incubator-hugegraph/pull/2289)
- migrate `hugegraph-hstore` module to `hugegraph-server` (also migrate `org.apache.hugegraph.meta` and `migrate org.apache.hugegraph.space` for compilation purpose)
- introduce `ISchemaTransaction` for adapting special version of schema tx for hstore backend
- query task and server info by graph tx for hstore backend
- modify index serialization and deserialization logic (merge index table) for hstore backend
- remove system dependency to hugegraph-core in hg-store-core

For detailed adaptation documentation, refer to https://hugegraph.feishu.cn/wiki/WdqXwyHkvi0UCvkBUDWcZgQHnhe

## Verifying these changes

- [ ] Trivial rework / code cleanup without any test coverage. (No Need)
- [ ] Already covered by existing tests, such as *(please modify tests here)*.
- [x] Need tests and can be verified as follows:
    - validate that the functionality of existing backends remains unaffected
    - verify the basic functionality of the hstore backend (basic graph import has been already verified)

| backend |                  test                   | status |
| :-----: | :-------------------------------------: | :----: |
| hstore  |           basic graph import            |   √    |
| hstore  | ldbc validation (without query 4 and 5) |   √    |
| rocksdb |           basic graph import            |   √    |
| rocksdb | ldbc validation (without query 4 and 5) |   √    |

## Does this PR potentially affect the following parts?

- [ ]  Nope
- [ ]  Dependencies (add/update license info) 
- [ ]  Modify configurations
- [ ]  The public API
- [x]  Other affects (maybe breaking changes)

## Documentation Status

- [x]  `Doc - TODO` 
- [ ]  `Doc - Done` 
- [ ]  `Doc - No Need`

### CODE REVIEW

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/StandardHugeGraph.java [hunk 1]
```diff
@@ -453,9 +458,25 @@ private void clearVertexCache() {
         }
     }
 
-    private SchemaTransaction openSchemaTransaction() throws HugeException {
+    private boolean isHstore() {
+        return this.storeProvider.isHstore();
+    }
+
+    private void initMetaManager() {
+        this.metaManager.connect("hg", MetaManager.MetaDriverType.PD,
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/StandardHugeGraph.java:466]: 这个链接是不是每个新的链接都要调用一次？是否只执行一次就可以呢？

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/auth/HugeRole.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+        public static final String ROLE = Hidden.hide("role");
+
+        public static final String ID = T.id.getAccessor();
+        public static final String LABEL = T.label.getAccessor();
+
+        public static final String NAME = "~role_name";
+        public static final String NICKNAME = "~role_nickname";
+        public static final String GRAPHSPACE = "~graphspace";
+        public static final String DESCRIPTION = "~role_description";
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/auth/HugeRole.java:194]: 这里咋不用hide, hide就是加～前缀吧

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/auth/SchemaDefine.java [hunk 3]
```diff
@@ -281,6 +290,19 @@ public abstract static class Relationship extends AuthElement {
 
         public abstract Id target();
 
+        public void setId() {
+            this.id(IdGenerator.of(this.source().asString() + "->" +
+                                   this.target().asString()));
+        }
+
+        public static <T extends Relationship> T fromMap(Map<String, Object> map, T entity) {
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/auth/SchemaDefine.java:298]: fromMap名字有点奇怪，要不要改一下？

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/ConditionQuery.java [hunk 4]
```diff
@@ -870,4 +876,14 @@ public interface ResultsFilter {
 
         boolean test(HugeElement element);
     }
+
+    public byte[] bytes() {
+        Gson gson = new GsonBuilder()
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/ConditionQuery.java:881]: Gson可以做成静态全局变量，不用每次创建了

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/serializer/QueryAdapter.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.reflect.TypeToken;
+
+public class QueryAdapter extends AbstractSerializerAdapter<Condition> {
+
+    static HashMap<String, Type> cls;
+
+    static {
+        cls = new HashMap() {{
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/serializer/QueryAdapter.java:43]: 可以用immutableMap嘛？

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/serializer/AbstractSerializerAdapter.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+
+public abstract class AbstractSerializerAdapter<T> implements JsonSerializer<T>,
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/serializer/AbstractSerializerAdapter.java:34]: 以后可以考虑优化成二进制协议

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/serializer/QueryIdAdapter.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.hugegraph.backend.id.Id;
+import org.apache.hugegraph.backend.id.IdGenerator;
+import org.apache.hugegraph.backend.serializer.BinaryBackendEntry;
+
+public class QueryIdAdapter extends AbstractSerializerAdapter<Id> {
+
+    static HashMap<String, Type> cls;
+
+    static {
+        cls = new HashMap() {{
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/serializer/QueryIdAdapter.java:34]: 同

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/store/BackendStoreInfo.java [hunk 8]
```diff
@@ -41,6 +41,9 @@ public boolean exists() {
     }
 
     public boolean checkVersion() {
+        if (this.storeProvider.isHstore()) { // skip for hstore now
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/store/BackendStoreInfo.java:44]: hstore暂时不用考虑兼容问题嘛？如果用户滚动升级，节点版本不一致，是否会有风险？
如果有的话，是否检查下版本比较合适呢？至少保证不兼容的版本不会产生脏数据

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+            Map<String, String> map = contents.getKvsMap();
+            for (Map.Entry<String, String> entry : map.entrySet()) {
+                System.out.println(entry.getKey() + ":" + entry.getValue());
+            }
+            System.out.println(map.size());
+        } catch (PDException e) {
+            e.printStackTrace();
+        }
+        Date dNow = new Date();
+        System.out.println(dNow);
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java:73]: log

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+                continue;
+            }
+            resultMap.put(key, value);
+        }
+        return resultMap;
+    }
+
+    @Override
+    public LockResult lock(String key, long ttl) {
+        while (true) {
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java:184]: 没有次数限制嘛？

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/DistributedLock.java [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+    private static ByteSequence toByteSequence(String content) {
+        return ByteSequence.from(content, Charset.defaultCharset());
+    }
+
+    public LockResult tryLock(String lockName, long ttl, long timeout) {
+        LockResult lockResult = new LockResult();
+        ScheduledExecutorService service =
+                Executors.newSingleThreadScheduledExecutor();
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/DistributedLock.java:68]: 这个是不是可以放全局队列来使用，没看错的话，这里的意思是没次lock就开一个线程来keep alive

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/PdDistributedLock.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+/**
+ * @author zhangyingjie
+ * @date 2022/6/18
+ **/
+public class PdDistributedLock extends AbstractDistributedLock {
+
+    private static int poolSize = 8;
+    private final KvClient client;
+    private ScheduledExecutorService service = new ScheduledThreadPoolExecutor(poolSize, r -> {
+        Thread t = new Thread(r);
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/PdDistributedLock.java:39]: 线程最好有个名字，比如这里keepalive-*

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/DistributedLock.java [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+
+        KeepAliveTask(Lease leaseClient, long leaseId) {
+            this.leaseClient = leaseClient;
+            this.leaseId = leaseId;
+        }
+
+        @Override
+        public void run() {
+            this.leaseClient.keepAliveOnce(this.leaseId);
+        }
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/DistributedLock.java:162]: 这里可以统计下调用的时间间隔

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/IdCounter.java [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+import org.apache.hugegraph.store.term.HgPair;
+import org.apache.hugegraph.type.HugeType;
+import org.apache.hugegraph.util.E;
+
+public class IdCounter {
+
+    private static final int TIMES = 10000;
+    private static final int DELTA = 10000;
+    private static final String DELIMITER = "/";
+    private static Map<String, HgPair> ids = new ConcurrentHashMap<>();
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/IdCounter.java:40]: volatile？

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/EtcdMetaDriver.java [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+    public boolean isLocked(String key) {
+        try {
+            long size = this.client.getKVClient().get(toByteSequence(key))
+                                   .get().getCount();
+
+            if (size > 0) {
+                return true;
+            } else {
+                return false;
+            }
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/EtcdMetaDriver.java:301]: return size>0

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+import org.apache.hugegraph.pd.grpc.kv.TTLResponse;
+import org.apache.hugegraph.pd.grpc.kv.WatchEvent;
+import org.apache.hugegraph.pd.grpc.kv.WatchResponse;
+import org.apache.hugegraph.pd.grpc.kv.WatchType;
+
+import com.google.common.base.Strings;
+
+public class PdMetaDriver implements MetaDriver {
+
+    KvClient<WatchResponse> client = null;
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java:47]: Private final

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+    PDClient pdClient = null;
+    private PdDistributedLock lock;
+
+    public PdMetaDriver(String pdPeer) {
+        PDConfig pdConfig = PDConfig.of(pdPeer);
+        this.client = new KvClient<>(pdConfig);
+        this.pdClient = PDClient.create(pdConfig);
+        lock = new PdDistributedLock(this.client);
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java:55]: This

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/AbstractDistributedLock.java [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * @author zhangyingjie
+ * @date 2022/6/18
+ **/
+public abstract class AbstractDistributedLock {
+
+    private static volatile AbstractDistributedLock defaultLock;
+
+    public static AbstractDistributedLock getInstance(AutoCloseable client) {
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/AbstractDistributedLock.java:32]: 这单例不合适吧？每次都要传client

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/PdDistributedLock.java [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+        try {
+            LockResponse response = this.client.lock(key, ttl);
+            boolean succeed = response.getSucceed();
+            LockResult result = new LockResult();
+            if (succeed) {
+                result.setLeaseId(response.getClientId());
+                result.lockSuccess(true);
+                long period = ttl - ttl / 4;
+                ScheduledFuture<?> future = service.scheduleAtFixedRate(() -> {
+                    synchronized (result) {
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/PdDistributedLock.java:60]: 这里为啥加锁

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/PdDistributedLock.java [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+    @Override
+    public void unLock(String key, LockResult lockResult) {
+        try {
+            LockResponse response = this.client.unlock(key);
+            boolean succeed = response.getSucceed();
+            if (succeed == false) {
+                throw new HugeException("Failed to unlock '%s' to pd", key);
+            }
+            if (lockResult.getFuture() != null) {
+                synchronized (lockResult) {
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/PdDistributedLock.java:81]: 同

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/space/SchemaTemplate.java [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+
+import com.google.common.collect.ImmutableMap;
+
+public class SchemaTemplate {
+
+    public static SimpleDateFormat FORMATTER = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+    protected Date createTime;
+    protected Date updateTime;
+    protected String creator;
+    private String name;
```
Comment (zyxxoo) [inline hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/space/SchemaTemplate.java:35]: final

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/auth/HugeRole.java (hunk 2 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/auth/HugeRole.java:194]: 已有的代码如 `HugeProject`，对于 property key 都是这样写的 🤔

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java (hunk 10 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java:184]: 这个方法注解为 Deprecated 且未被引用，直接删除？

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/ConditionQuery.java (hunk 4 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/ConditionQuery.java:893]: done

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/serializer/QueryAdapter.java (hunk 5 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/serializer/QueryAdapter.java:43]: done

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java (hunk 9 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java:73]: 临时代码，已删除

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/DistributedLock.java (hunk 11 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/DistributedLock.java:68]: 参考 PdDistributedLock 进行了修改

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/PdDistributedLock.java (hunk 12 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/PdDistributedLock.java:39]: done

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/EtcdMetaDriver.java (hunk 15 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/EtcdMetaDriver.java:301]: done

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java (hunk 16 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java:47]: done

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java (hunk 17 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java:55]: done

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/AbstractDistributedLock.java (hunk 18 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/lock/AbstractDistributedLock.java:32]: 未被引用的抽象类，已删除

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/space/SchemaTemplate.java (hunk 21 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/space/SchemaTemplate.java:35]: done

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/serializer/QueryIdAdapter.java (hunk 7 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/serializer/QueryIdAdapter.java]: done

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/store/BackendStoreInfo.java (hunk 8 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/store/BackendStoreInfo.java:44]: 此处跳过是因为 hstore 不提供 system store，考虑从 hstore 的 graph store 中获取 driver version

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/StandardHugeGraph.java (hunk 1 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/StandardHugeGraph.java:470]: 考虑在 StandardHugeGraph 构造函数中初始化 meta manager

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/serializer/AbstractSerializerAdapter.java (hunk 6 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/query/serializer/AbstractSerializerAdapter.java:35]: 已添加注释

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/cache/CachedSchemaTransactionV2.java [hunk 22]
```diff
@@ -0,0 +1,469 @@
...
+
+        @PerfUtil.Watched
+        public V get(HugeType type, Id id) {
+            assert id.number();
+            long longId = id.asLong();
+            if (longId <= 0L) {
+                assert false : id;
+                return null;
+            }
+            int key = (int) longId;
```
[the remark this answers is not in the document]
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/cache/CachedSchemaTransactionV2.java:372]: @imbajin ignored? The existing `CachedSchemaTransaction` also has the same issue.

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java (hunk 10 above)
Comment (zyxxoo) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/meta/PdMetaDriver.java]: 如果是 deprecated，就直接删除吧

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/IdCounter.java (hunk 14 above)
Comment (VGalaxies) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/IdCounter.java]: should be `final`

File: hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/cache/CachedSchemaTransactionV2.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (imbajin) [reply hugegraph-server/hugegraph-core/src/main/java/org/apache/hugegraph/backend/cache/CachedSchemaTransactionV2.java:372]: > @imbajin ignored? The existing `CachedSchemaTransaction` also has the same issue.

Fine, both OK to ignore or fix it

**Pull request merged**